elastic/kibana
# [Security solution] AWS Bedrock connector

## Summary

New Generative AI connector to support AWS Bedrock. This change will need name change follow up as described here: https://github.com/elastic/kibana/issues/166960

Implemented with the sub actions framework. The actions are `runApi` and `invokeAI`. `runApi` is the action attached to the `test` sub action. `runApi` is responsible for making a `POST` request to the Bedrock API endpoint and returning the response data. This is an example payload of what Bedrock is expecting:
```
{
  "prompt": "\n\Human: Hello world! \n\nAssistant:",
  "max_tokens_to_sample": 300,
  "stop_sequences": ["\n\nAssistant:"],
}
```

### Create connector
<img width="1452" alt="Screenshot 2023-09-21 at 4 05 48 PM" src="https://github.com/elastic/kibana/assets/6935300/b66be1a7-877d-4bec-a331-ec2ab3dc9b8e">
<img width="1453" alt="Screenshot 2023-09-21 at 4 06 16 PM" src="https://github.com/elastic/kibana/assets/6935300/6cb35b81-7a14-4498-9e1b-2daa93c19708">

### Test connector
<img width="1310" alt="Screenshot 2023-09-21 at 4 07 24 PM" src="https://github.com/elastic/kibana/assets/6935300/4f2977c5-4c6a-4260-9235-f4db453e270f">


## Testing for ResponseOps team
My apologies for doing this all in one PR. Hopefully clear testing expectations will win me points back?
[Bedrock credentials](https://p.elstc.co/paste/rxa+OUI2#Nzv+-vFnsJB7xIkzb18rdVgNAJVTmYLZHsrKcHmt02N)
ResponseOps should test that a connector can be created and tested from the Stack Management > Connectors page

1. Create Bedrock connectors with [Bedrock credentials](https://p.elstc.co/paste/rxa+OUI2#Nzv+-vFnsJB7xIkzb18rdVgNAJVTmYLZHsrKcHmt02N)
2. Test the connector by sending example payload
3. Confirm results in Network tab. Should look like:
    <img width="506" alt="Screenshot 2023-09-21 at 4 10 47 PM" src="https://github.com/elastic/kibana/assets/6935300/0c6437e8-a4b3-4f3e-9a38-04e2b8a1c19d">


## Testing for Security Solution team
[Bedrock credentials](https://p.elstc.co/paste/rxa+OUI2#Nzv+-vFnsJB7xIkzb18rdVgNAJVTmYLZHsrKcHmt02N)
Security Solution should test that a connectors can be created 3 places within the AI assistant, that multiple connectors can be used in a single conversation, and that the connectors work with both Langchain and non-langchain:

1. Start fresh with no connectors and a cleared local storage (no conversation history)
2. Through the welcome conversation, create a Bedrock connector using the button in the conversation:
<img width="1309" alt="Screenshot 2023-09-21 at 4 14 15 PM" src="https://github.com/elastic/kibana/assets/6935300/a651a186-69bc-4fb7-a59f-7fef79b61d1b">
<img width="1307" alt="Screenshot 2023-09-21 at 4 14 25 PM" src="https://github.com/elastic/kibana/assets/6935300/18d4fcf4-a82e-4bac-a03c-48cba5f98fb8">
<img width="1309" alt="Screenshot 2023-09-21 at 4 14 59 PM" src="https://github.com/elastic/kibana/assets/6935300/05cbca30-468c-42b2-8fe2-7bbf3f02c783">

3. Send a message to confirm the connection
4. Create another Bedrock connector from the top left connector selector
<img width="1308" alt="Screenshot 2023-09-21 at 4 17 14 PM" src="https://github.com/elastic/kibana/assets/6935300/f019a053-ca36-4081-b189-601835d23d8f">

5. Create an OpenAI/GenA connector from the settings 
<img width="1309" alt="Screenshot 2023-09-21 at 4 16 10 PM" src="https://github.com/elastic/kibana/assets/6935300/54ca1381-2efb-4f7b-8097-07d5ea8fbc08">

6. Ensure that the connectors are listed with the proper description:
<img width="1339" alt="Screenshot 2023-09-20 at 11 55 31 AM" src="https://github.com/elastic/kibana/assets/6935300/f92e0609-3628-4a3b-879a-ca375a7f08fc">

7. Switch between connectors and have a chat with each, ensure that we can switch between Bedrock and OpenAI without issues
8. If you've had the setting `assistantLangChain={false}`, switch it to true or vice versa. You do not need to test the connector creation with langchain and non-langchain, but test sending messages with both langchain and non-langchain to both connector types
